src/pairingheap.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * Datastructures: GAP package providing common datastructures.
+ * Licensed under the GPL 2 or later.
+ *
+ * Implementation of a binary heap.
+ *
+ * Binary heaps are of course pretty standard data structures.
+ * However, a few design choices are possible. The implementation
+ * below has been influenced by this StackOverflow answer:
+ *   <https://stackoverflow.com/questions/6531543>
+ */
+
+#include "src/compiled.h"
+#include "pairingheap.h"
+
+#define DS_BINARYHEAP_ISLESS(heap) ELM_PLIST(heap, 1)
+#define DS_BINARYHEAP_DATA(heap) ELM_PLIST(heap, 2)
+
+
+static Obj SumSmallInt(Obj a, Obj b)
+{
+    GAP_ASSERT(IS_INTOBJ(a));
+    GAP_ASSERT(IS_INTOBJ(b));
+    Obj sum;
+    if (!SUM_INTOBJS(sum, a, b))
+        ErrorMayQuit("PANIC: counter overflow in pairing heap code", 0, 0);
+    return sum;
+}
+
```

Comment:
The difference is that it Errors out on overflow. There are probably lots of places in the kernel where SUM_INTOBJS is used and the code just silently assumes now overflow. 

A version that merely used GAP_ASSERT to check for lack of overflow in debug mode might also make sense.